File: Stack.ts
type StackNode<T> = {
  value: T;
  previus?: StackNode<T>;
};

export default class Stack<T> {
  public length: number;
  private head?: StackNode<T>;

  constructor() {
    this.head = undefined;
    this.length = 0;
  }

  push(item: T) {
    const node = { value: item } as StackNode<T>;
    this.length++;

    if (!this.head) {
      this.head = node;
      return;
    }
    node.previus = this.head;
    this.head = node;
  }

  pop() {
    if (this.length === 0) {
      const head = this.head;
      this.head = undefined;
      return head?.value;
    }

    this.length--;
    const head = this.head as StackNode<T>;
    this.head = head.previus;
    return head.value;
  }

  peek() {
    return this.head?.value;
  }
}

File: Stack.py
class Node:
    def __init__(self,data):
        self.value = data;
        self.prev = None

class Stack:

    def __init__(self):
        self.length = 0
        self.head = None

    def push(value):
        if self.length == 0:
            self.length += 1
            self.head = Node(value)
            return
        
        newNode = Node(value)
        newNode.prev = self.head
        self.head = newNode

    def pop():
        if self.length == 0:
            head = self.head
            self.head = None
            return  head.value if head else None

        self.length -= 1
        head = sef.head
        self.head = self.head.prev
        return head.value 

    def peek():
        return self.head.value if self.head else None
        

